llarp/link/link_manager.cpp

```diff
@@ -1062,14 +1062,178 @@ namespace llarp
+
+      auto hop_info = frame_dict.require<ustring>("encrypted");
+      TunnelNonce nonce;
+      nonce.from_string_view(frame_dict.require<std::string_view>("nonce"));
+      Pubkey otherPubkey;
+      otherPubkey.from_string_view(frame_dict.require<std::string_view>("pubkey"));
+
+      auto crypto = CryptoManager::instance();
+
+      SharedSecret shared;
+      // derive shared secret using ephemeral pubkey and our secret key (and nonce)
+      if (!crypto->dh_server(shared, otherPubkey, router.identity(), nonce))
+      {
+        log::info("DH failed during path build.");
+        m.respond(serialize_response({{"STATUS", PathBuildMessage::BAD_CRYPTO}}), true);
+        return;
+      }
+
+      // hash data and check against given hash
+      ShortHash digest;
+      if (!crypto->hmac(
+              digest.data(),
+              reinterpret_cast<unsigned char*>(frame_str.data()),
+              frame_str.size(),
+              shared))
+      {
+        log::error(link_cat, "HMAC failed on path build request");
+        m.respond(serialize_response({{"STATUS", PathBuildMessage::BAD_CRYPTO}}), true);
+        return;
+      }
+      if (!std::equal(
+              digest.begin(), digest.end(), reinterpret_cast<const unsigned char*>(hash.data())))
+      {
+        log::info("HMAC mismatch on path build request");
+        m.respond(serialize_response({{"STATUS", PathBuildMessage::BAD_CRYPTO}}), true);
+        return;
+      }
+
+      // decrypt frame with our hop info
+      if (!crypto->xchacha20(
+              reinterpret_cast<unsigned char*>(hop_info.data()), hop_info.size(), shared, nonce))
+      {
+        log::info("decrypt failed on path build request");
+        m.respond(serialize_response({{"STATUS", PathBuildMessage::BAD_CRYPTO}}), true);
+        return;
+      }
+
+      // populate transit hop object with hop info
+      // TODO: how to get downstream hop RouterID from here (all we have is oxen::quic::message)
+      //       could do message->btstream->stream->connection_interface->connectionid
+      //       and check our mapping, but that feels ugly as sin (and message->stream is private)
+      // TODO: also need downstream for IP / path build limiting clients
+      auto hop = std::make_shared<path::TransitHop>();
+      // hop->info.downstream = m.from(); // TODO: RouterID m.from() or similar
+      auto hop_dict = oxenc::bt_dict_consumer{hop_info};
+
+      // extract pathIDs and check if zero or used
+      hop->info.txID.from_string_view(hop_dict.require<std::string_view>("txid"));
+      hop->info.rxID.from_string_view(hop_dict.require<std::string_view>("rxid"));
+      if (info.txID.IsZero() || info.rxID.IsZero())
+      {
+        log::info("Invalid PathID; PathIDs must be non-zero.");
+        m.respond(serialize_response({{"STATUS", PathBuildMessage::BAD_PATHID}}), true);
+        return;
+      }
+      hop->info.upstream.from_string_view(hop_dict.require<std::string_view>("next");
```

Comment:
Could maybe just do:
`const auto& hop_info = hop->info`

Then every call is just 
`hop_info.txid` or `hop_info.rxid`, etc